### ISSUE TITLE
Device uuid persistence across config changes

### ISSUE THREAD

**Issue by justjam2013** [homebridge/homebridge#3712]
### Current Situation

I don't know the exact behavior under the hood, so the following is my assumption based on observed behavior.
When the configuration of an existing device changes (for example name change or delay length) HB deletes the existing device and creates a new one, so a new unique device identifier is created. This means that in Homekit the device is deleted, along with all scenes and automations that only use this device and all automations that include this device.
This means that if I want to change to change a device name, for example the name of a switch from "Play greeting switch" to "Play welcome home greeting", so that I can support additional behavior, for example another switch called "Play good morning greeting switch", I either have to recreate all the scenes and automations, or I keep a "Temp switch". I replace "Play greeting switch" with "Temp switch" everywhere, rename the switch (which deletes switch "Play greeting switch" and creates new switch "Play welcome home greeting"), then replace "Temp switch" with "Play welcome home greeting".
It would be nice if the uuid of the device survived the name change, along with all of my Homekit scenes and automations.

### Proposed Change

The Homebridge UI automatically adds a UUID field for each device to the configuration form (maybe a hidden or a readonly field) and stores it in the config json. So if the user changes a property for an existing device, the uuid survives the changes. The new uuid would be generated only if the user specifically deletes a particular device in the UI config form and creates a new device.

**The user might use the JSON Config screen to modify the configuration.**

1. If the user modifies device properties, but not the UUID, then the device is not deleted and recreated.
2. If the user modifies the UUID, then the old device is deleted and a new device is created.
3. If the user creates new devices, missing UUIDs are programmatically added when the user saves the configuration.

**The user might modify the config json by hand in the filesystem.**

1. When Homebridge starts up, it checks the configuration and adds missing UUIDs programmatically and updates the configuration.

### Additional Context

Because of the multiple routes to updating the Homebridge configuration, this is a change to both Homebridge and Homebridge UI.
Link to Homebridge UI issue: [Device uuid persistence across config changes #2201](https://github.com/homebridge/homebridge-config-ui-x/issues/2201)

**Comment by NorthernMan54** [comment]
No need to create 2 issues, as we can transfer them between repo, or link fixes across repos, so pls close one of them.

The creation and deletion of accessories is controlled by the plugin and not homebridge or the UI.  All homebridge does is provide the interface to create the plugin.  See https://developers.homebridge.io/#/api/platform-plugins

sounds like the plugin you are using here, recreates the accessory every time it starts.  And potentially using name as part of the uuid

**Comment by justjam2013** [comment]
Closed Homebridge UI issue.

**Comment by justjam2013** [comment]
You are correct that the plugins use the accessory name. According to the code, the plugin must use a unique identifier to register the accessory. However the plugin does not store any state. The state is maintained by the config that is passed back to the plugin every time it is started up.
So when the accessory config looks like this:

    {
        "name": "Delay Switch 1",
        "delay": "1800",
        "other_config_value": "xyz"
    },
    {
        "name": "Delay Switch 2",
        "delay": "3600",
        "other_config_value": "abc"
    },

The only place the plugin can get a UUID is by 1) using the name value, 2) the name + combination of other field values, 3) some unique value like timestamp in milliseconds.
A plugin will provide a different UUID every time around, unless it itself either implements a UUID field in the configuration or stores a static value stored in the plugin. But this is something that only a plugin that creates a single accessory can reasonably do.
So, yes, creation and deletion of accessories is controlled by the plugin. Ideally, the plugin should create a durable UUID by storing it in the config. Without an immutable UUID, the plugin cannot return the same value back to Homebridge to indicate an update to an accessory, and Homebridge has no way of knowing if the plugin just meant to update an accessory or meant to delete it and create a new one.

But I would argue that Homebridge is an accomplice in enabling bad coding practices. If Homebridge automatically generated durable and immutable UUIDs for each accessory, they would be passed back to the plugin in the config. Now the plugin would be responsible in deciding whether to pass it back to Homebridge, indicating the same device was updated, or pass back a new UUID indicating that the device was deleted and a new one was created.

Homebridge already keeps a list of registered accessories and checks for duplicate registration ids in `plugin.ts:114`:

    public registerAccessory(name: AccessoryName, constructor: AccessoryPluginConstructor): void {
        if (this.registeredAccessories.has(name)) {
            throw new Error(`Plugin '${this.getPluginIdentifier()}' tried to register an accessory '${name}' which has already been registered!`);
        }
        ...
        this.registeredAccessories.set(name, constructor);
    }

So my suggestion is to remove the UUID management from plugins (most of which, if not all, are not doing it properly) and manage it within Homebridge.

This is akin to a database auto generating a UUID for a record.

**Comment by NorthernMan54** [comment]
This ultimately is a plugin issue, as the plugin isn't using the recommended approach of using [dynamic platform type](https://developers.homebridge.io/#/).

PS You have not indicated which plugin you are experiencing this with ?

**Comment by justjam2013** [comment]
I looked at [dynamic platform type](https://developers.homebridge.io/#/) and it doesn't say anything about UUIDs.

In `platform.ts:72-75` the comment specifies the requirements for the UUID, but I had to go dig in the code:

      // generate a unique id for the accessory this should be generated from
      // something globally unique, but constant, for example, the device serial
      // number or MAC address
      const uuid = this.api.hap.uuid.generate(device.exampleUniqueId);

MAC address is only useful if the hardware device has a MAC. If it's a virtual device running on Homebridge, then all will share the MAC of the Homebridge host.
In the case of a plugin creating multiple accessories, we are now back to creating or storing multiple unique device serial numbers. The serial number is likely generated using the accessory name, so it is durable, but not immutable.

So yes, I am not debating that ultimately the plugin author is responsible for implementing a bad UUID. I'm not arguing that. I'm just saying that a number of plugins likely suffer from this problem. And I'm suggesting to move control over this to Homebridge.

The two plugins that I have experienced this with are
- homebridge-homepod-radio (supports multple radio accessories and multiple switch accessories):

          const uuid = hap.uuid.generate('homebridge:homepod:radio:' + radio.name);
          const uuid = hap.uuid.generate('homebridge:homepod:fileSwitch:' + fileSwitch.name);

The last update is 6 months ago and the last response to an opened issue is 3 months ago. Also, the author of this plugin is in Ukraine, so I'm sure that updating this plugin is not top of his priorities right now.

- homebridge-dummy, which is a verified plugin, but doesn't implement the dynamic template. It looks like it implements the now deprecated accessory plugin. Probably the best solution, but also the most time consuming, would be to rewrite this plugin with the dynamic template.

Moving UUID management into Homebridge means that these plugins could be patched with minimal code changes.

**Comment by donavanbecker** [comment]
PRs are welcomed!

**Comment by justjam2013** [comment]
@donavanbecker I think that @NorthernMan54 is making good points. I take it as trying to clarify if there is an issue, what the issue is, where does the issue originate from, is it something that makes sense to address in Homebridge. All valid questions. I am submitting the enhancement suggestion, so I believe that the responsibility to do all the work of investigating and answering those questions is on me.

**Comment by hjdhjd** [comment]
I don't think this belongs in Homebridge. UUIDs are the responsibility, ultimately, of the plugin not of Homebridge. Agree that documentation can always be improved, but critically: **UUID management is the explicit, contractual responsibility of plugins**. Even more so the case with platform plugins.

Sure - some people use MAC addresses, because they're generally unique. When MAC addresses can't be used the **plugin developer** has the responsibility to decide what to use. Management of UUIDs should most definitely not migrate into Homebridge - it's the obligation of the plugin to manage.

TL;DR: it is expressly the responsibility of the plugin author to ensure UUIDs are unique within the context and namespace of their plugin.

PS: UUIDs do not need to be unique **across** plugins. UUID uniqueness is specific to the plugin. This is part of why it's the contractual obligation of the plugin and not something that belongs in Homebridge proper. If Homebridge needed to manage UUIDs holistically, I can see an argument here - but it does not, by design.

**Comment by justjam2013** [comment]
It's not about UUID uniqueness. It's about UUID durability across all plugin restarts.
The fact that the UUID value is generated by the plugin each time around, means that the UUID is not durable. For example, the author makes code changes to the UUID creation. That is fine going forward for new accessories. But this breaks all accessories that were created with the old UUID code, and they would not survive being deleted and recreated.
The issue is that the UUID is passed from the plugin to Homebridge, but not from Homebridge back to the plugin.
It's like telling the browser to recreate the session token every single time it sends a request to the server. Once the session token is created, it is passed back and forth until the token is renegotiated.

A plugin that creates a single accessory can have a hardcoded UUIDs.
Plugins that create multiple accessories dynamically, like switches, cannot hardcode UUIDs and UUIDs are always generated with a random value. So the only way those plugins can create UUIDs that are durable are:

1. Store the UUID in the name field along with the actual name: "My Accessory - 203b7a1c-279d-49e8-803a-e93b93358a2a"
2. Create a uuid  field in the config: "uuid" : "203b7a1c-279d-49e8-803a-e93b93358a2a"
3. Maintain a local dictionary: "name" : "203b7a1c-279d-49e8-803a-e93b93358a2a"

The moment the user manually changes the name of the accessory in the json, option 3 is broken. Same if the user creates the config manually.
Option 1 breaks if a user makes a change to the name in the json, not understanding the importance of that weird hex string. And in all fairness, the UUID is not to a name, so it should be in it's own field. That is option 2.

So to create a durable UUID, each plugin author needs to add a uuid field to their config and store a UUID value.
Or Homebridge could provide that boilerplate functionality out of the box:

1. When the plugin registers an accessory, a UUID field will be created in the configuration
2. If the plugin does not provide a UUID, one will be autogenerated for it and stored in the configuration
3. When the plugin receives the configuration back when it starts up, it receives the original token it sent when it registered that accessory
4. When the plugin wants to update an accessory, it passes back the original token it sent when it registered that accessory

**> UUID management is the explicit, contractual responsibility of plugins**
I'm not arguing the contractual responsibility of plugins to manage their own UUIDs. All I'm saying is that UUIDs should be managed like a persistent session token and be exchanged between Homebridge and plugins, until it is explicitly renegotiated.

**Comment by hjdhjd** [comment]
I do understand the crux of the issue - I remain unmoved. 

The core problem isn't Homebridge, it's plugins that need to be modernized to be well-behaved in the modern Homebridge world. We've got a philosophical disagreement perhaps on where this belongs - I don't agree that this belongs in Homebridge. The proper fix here is to modernize your plugins.

**All** of my plugins, for instance, create a multitude of dynamic accessories/services. They do not run into these issues, survive name changes just fine, as do lots of other plugins.

You're advocating for a solution to misbehaving or not-well-designed plugins - that's my core dissent here. Badly written plugins shouldn't get additional accommodation.

And no - UUID management works well as it is right now and I have a high bar for moving more things into Homebridge unless absolutely necessary. This doesn't meet that standard, in my mind.

**Comment by justjam2013** [comment]
> Badly written plugins shouldn't get additional accommodation.

**Comment by hjdhjd** [comment]
@justjam2013 Thank you for a thoughtful discussion…always appreciated and embraced!
